pkg/cloudprovider/cloudprovider.go

```diff
@@ -27,127 +27,139 @@ type Client struct {
+	configFile string
 }
 
 // ClientInt client interface
 type ClientInt interface {
 	UpdateUserMSI(addUserAssignedMSIIDs, removeUserAssignedMSIIDs []string, name string, isvmss bool) error
 	GetUserMSIs(name string, isvmss bool) ([]string, error)
+	Init() error
 }
 
 // NewCloudProvider returns a azure cloud provider client
 func NewCloudProvider(configFile string) (c *Client, e error) {
-	azureConfig := config.AzureConfig{}
-	if configFile != "" {
+	client := &Client{
+		configFile: configFile,
+	}
+
+	if err := client.Init(); err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func withInspection() autorest.PrepareDecorator {
+	return func(p autorest.Preparer) autorest.Preparer {
+		return autorest.PreparerFunc(func(r *http.Request) (*http.Request, error) {
+			klog.Infof("Inspecting Request: Method: %s \n URL: %s, URI: %s\n", r.Method, r.URL, r.RequestURI)
+
+			return p.Prepare(r)
+		})
+	}
+}
+
+// Init initializes the cloud provider client based
+// on a config path or environment variables
+func (c *Client) Init() error {
+	c.Config = config.AzureConfig{}
+	if c.configFile != "" {
 		klog.V(6).Info("Populate AzureConfig from azure.json")
-		bytes, err := ioutil.ReadFile(configFile)
+		bytes, err := ioutil.ReadFile(c.configFile)
 		if err != nil {
-			klog.Errorf("Read file (%s) error: %+v", configFile, err)
-			return nil, err
+			klog.Errorf("Read file (%s) error: %+v", c.configFile, err)
+			return err
 		}
-		if err = yaml.Unmarshal(bytes, &azureConfig); err != nil {
+		if err = yaml.Unmarshal(bytes, &c.Config); err != nil {
 			klog.Errorf("Unmarshall error: %v", err)
-			return nil, err
+			return err
 		}
 	} else {
 		klog.V(6).Info("Populate AzureConfig from secret/environment variables")
-		azureConfig.Cloud = os.Getenv("CLOUD")
-		azureConfig.TenantID = os.Getenv("TENANT_ID")
-		azureConfig.ClientID = os.Getenv("CLIENT_ID")
-		azureConfig.ClientSecret = os.Getenv("CLIENT_SECRET")
-		azureConfig.SubscriptionID = os.Getenv("SUBSCRIPTION_ID")
-		azureConfig.ResourceGroupName = os.Getenv("RESOURCE_GROUP")
-		azureConfig.VMType = os.Getenv("VM_TYPE")
-		azureConfig.UseManagedIdentityExtension = strings.EqualFold(os.Getenv("USE_MSI"), "True")
-		azureConfig.UserAssignedIdentityID = os.Getenv("USER_ASSIGNED_MSI_CLIENT_ID")
+		c.Config.Cloud = os.Getenv("CLOUD")
+		c.Config.TenantID = os.Getenv("TENANT_ID")
+		c.Config.ClientID = os.Getenv("CLIENT_ID")
+		c.Config.ClientSecret = os.Getenv("CLIENT_SECRET")
+		c.Config.SubscriptionID = os.Getenv("SUBSCRIPTION_ID")
+		c.Config.ResourceGroupName = os.Getenv("RESOURCE_GROUP")
+		c.Config.VMType = os.Getenv("VM_TYPE")
+		c.Config.UseManagedIdentityExtension = strings.EqualFold(os.Getenv("USE_MSI"), "True")
+		c.Config.UserAssignedIdentityID = os.Getenv("USER_ASSIGNED_MSI_CLIENT_ID")
 	}
 
-	azureEnv, err := azure.EnvironmentFromName(azureConfig.Cloud)
+	azureEnv, err := azure.EnvironmentFromName(c.Config.Cloud)
 	if err != nil {
 		klog.Errorf("Get cloud env error: %+v", err)
```

Comment:
We should probably just wrap the error with a message and return instead of logging here. Not in the scope of this PR, but we should probably do a follow up.

Reply:
Opened an issue for this: https://github.com/Azure/aad-pod-identity/issues/604.